File: 1_2.cpp
#include <iostream>
#include <cmath>
#include <cstdlib>
#include <ctime>

using namespace std;

double func(double x) {
    return pow(sin(x), 2) / (13 - 12 * cos(x));
}

double monte_carlo_integration(double (*f)(double), double a, double b, int n) {
    double sum = 0.0;
    double x, y;

    srand(time(NULL));

    for (int i = 0; i < n; i++) {
        x = a + (double)rand() / RAND_MAX * (b - a);
        y = (double)rand() / RAND_MAX;

        if (y <= f(x)) {
            sum += 1;
        }
    }

    return sum / n * (b - a);
}

int main() {
    double a, b;
    int n;

    cout << "Введите нижний предел интегрирования: ";
    cin >> a;
    cout << "Введите верхний предел интегрирования: ";
    cin >> b;
    cout << "Введите количество случайных точек: ";
    cin >> n;

    double integral = monte_carlo_integration(func, a, b, n);
    cout << "Приближенное значение интеграла: " << integral << endl;

    return 0;
}

File: 1_2.py
import math
import random

def func(x):
    return math.sin(x) ** 2 / (13 - 12 * math.cos(x))


def monte_carlo_integration(f, a, b, n):
    sum_hits = 0.0

    # Опционально установить начальное значение генератора случайных чисел для повторяемости
    random.seed()

    for _ in range(n):
        x = a + random.random() * (b - a)
        y = random.random()

        if y <= f(x):
            sum_hits += 1

    return sum_hits / n * (b - a)


def main():
    a = float(input("Введите нижний предел интегрирования: "))
    b = float(input("Введите верхний предел интегрирования: "))
    n = int(input("Введите количество случайных точек: "))

    integral = monte_carlo_integration(func, a, b, n)
    print(f"Приближенное значение интеграла: {integral}")


if __name__ == "__main__":
    main()
